implementation-details.md

```diff
@@ -284,32 +313,53 @@ The `CrdsFilter`s are partitioned in pull requests between peers selected random
+The higher the node's weight, the more pull requests will be sent to it.
 
-Each pull request contains one of the `CrdsFilter`s from the list and a `CrdsValue`, which is usually a `LegacyContactInfo`. It contains details of the node that sent the pull request - a list of its IP addresses, public key and wallclock. 
+Each pull request contains one of the `CrdsFilter`s from the list and a `CrdsValue`, which is a `LegacyContactInfo`. It contains details of the node that sent the pull request - a list of its IP addresses, public key and wallclock. 
 
-When a node receives a pull request it inserts the pull request value into its `crds`, or updates the existing one in case of duplicate. Then, it checks whether the pull request comes from a valid address as a pull response needs to be sent back. Next, the node checks the pull request wallclock value - if it's not within 15 seconds of the current node's wallclock the request is ignored. Otherwise, a node gathers data from the `crds` table and filters it with the provided `CrdsFilter`. Node uses the `mask_bits` value and the `CrdsShards` structure to find the proper hashes of `CrdsValue`s from `crds`:
-1. If `mask_bits = shard_bits = 12` the `CrdsFilter` `mask` value can be used directly as an index in the `shards` vector. The node will gather all `CrdsValue`s from `crds` whose hashes belong to that index in `shards` and filter them with the provided bloom filter. Values not existing in the filter will be collected for further processing.
-> _Example_
->
-> If `mask = 000000010000 = 16` we search for hashes from `shards` vector under index 16
+When a node receives a pull request it inserts the pull request value, `LegacyContactInfo`, into its `crds`, or updates the existing one in case of duplicate. Then, it checks whether the pull request comes from a valid address as a pull response needs to be sent back. Next, the node checks the pull request wallclock value - if it's not within 15 seconds of the current node's wallclock the request is ignored. Otherwise, a node gathers data from the `crds` table and filters it with the provided `CrdsFilter`. 
 
-2. If `mask_bits < shard_bits` values are gathered from all `shards` where first `mask_bits` bits of the indices are equal to the `mask` and then they are filtered using the bloom filter.
-> _Example_
->
-> If `mask_bits = 4`, `shard_bits = 12` and `mask = 0001` we search for all shards which contain hashes starting with `0001`, e.g. `000100000000...`, `000100000001...`, and so on.
+As was already [explained](#crds), the `CrdsShards` structure holds a `shards` vector whose elements are an index map of `CrdsValue`s indexes in the `crds` table and their hashes. The `shards` vector allows the node to quickly find and filter out `CrdsValue`s that the pull request sender already has in its `crds`. Let's assume `shard_bits = 3` (actually it is 12), `mask_bits = 3` and `mask = 001 = 1` for simplicity. The `shards` vector would look like this:
```

Comment:
> Let's assume `shard_bits = 3` (actually it is 12)

can change to: For simplicity, let's assume `shard_bits = 3` (In reality it is 12)...